### PR TITLE
PoC: the Liquid Salt system

### PR DESCRIPTION
**Pull request by mirceaulinic**
We have discussed some time ago under #41879 regarding the possibility of a pluggable system to allow Salt to load (further) configuration options for various systems.

This system can have the advantage of managing the configuration opts (or some of them) from a central point, without having to regenerate the config files on hundreds/thousands of Minions. Of course, this comes with the risk that with a single click you can break your entire infrastructure. But it's the same risk when using the `ext_pillar`, so the user assumes the risk (we'll probably need to document that).

That can be a very sensitive subject, as @whiteinge mentioned:

> Managing the configs from a GUI or automating deploying config changes is a bit tricker -- since the daemon will need to be restarted to pick up any changes we need to be certain in advance that the daemon will restart successfully. Suggestions have been to perform a pre-check validation step (#4352) or to fail gracefully. But if there is a problem in the generated config such as bad syntax or incorrect parameters it can cause the master to not start or to traceback.

What I have so far is very far away from achieving these goals, for the moment I only want to present what I currently have, and discuss if it makes sense to you to continue on this path.

The base idea is that you can put at least (or only) the following lines in your Master / Minion config file and the rest of the options are fetched from the systems you want, e.g.,:

```/etc/salt/master```
```yaml
liquid:
  - yaml:
      path: https://example.com/yaml-api
  - mongo:
      host: example.com
      port: 49017
      db: common_opts
  - mongo:
      host: example.com
      port: 49017
      db: master_only_opts
```

I have called this system _Liquid_ (because https://en.wikipedia.org/wiki/Molten_salt_reactor#Liquid_salt_very_high_temperature_reactor), but maybe _Molten_ could be a better name? (Molten-salt fueling options). Let me know if we should rename. :-)

### Design choices

- You still need to provide the `/etc/salt/master` and `/etc/salt/minion` files, that have at least the liquid config.
- The opts collected via `liquid` will override anything else.
- The opts collected via `liquid` will be cached in a file called `_liquid.conf` under `/etc/salt/minion.d/_liquid.conf` (or `/etc/salt/master.d/_liquid.conf`). Is this a good enough place? I preferred so, because they are collected from there anyway, but it has the disadvantage that the user may turn off the liquid system and the opts will continue to be loaded from there. Probably we should save it elsewhere? If so, where?

#### What works so far:

- A pluggable system.
- Loads the opts from the sources configured.
- An execution module for CLI manipulations.

#### Challenges:

- During startup the configuration opts from the `/etc/salt/master` and `/etc/salt/minion` files are read several times (I added a error log just before this line: https://github.com/saltstack/salt/blob/develop/salt/config/__init__.py#L2140)

```
$ sudo salt-master -l error
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /root/.saltrc
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
[ERROR   ] Read config from /etc/salt/master
...
... snip ...
...
```

We probably should avoid hitting the remote services / APIs so much. However I added a `liquid_always_fetch` option that can be turned on if the user really wants that.

But I don't know how can I detect that the Minion/Master process is just starting up, and not loading the opts the 100-th time already? Is there any flag I missed? I added an artificial `__liquid_loaded` in the opts, but that turns to be a very weak way. Help please! :-)

If you like the idea and think there's value to continue this system, please let me know (and if possible help with the question(s) above), then I will finalize and document.

#### How to test this branch?

Just add the following in the Master or Minion config file:

```yaml
liquid:
  - yaml: {}
```

And you should see the following option loaded (e.g., using `test.get_opts`):

```yaml
some_test_data_from_the_liquid_yaml: True
```

Ping @cachedout @rallytime @garethgreenaway (from #41879)

### CODE REVIEW

File: salt/modules/saltutil.py [hunk 1]
```diff
@@ -623,6 +623,44 @@ def sync_thorium(saltenv=None, refresh=False, extmod_whitelist=None, extmod_blac
     return ret
 
 
+def sync_liquid(saltenv=None, refresh=False, extmod_whitelist=None, extmod_blacklist=None):
+    '''
+    .. versionadded:: Oxygen
```
Comment (mirceaulinic) [inline salt/modules/saltutil.py]: Note: I don't necessarily aim to get this into Oxygen. But if I can finish before the freeze, it may be nice to have.

**Comment by rallytime** [comment]
Thanks @mirceaulinic!

I'd like to get @cachedout's feed back on this. :)

**Review by cachedout — APPROVED** [review]
This covers all my main concerns and is very nicely implemented.

**Comment by thatch45** [comment]
I need to apologise for being so slow here, but this is beyond brilliant. I also think it would make a lot of sense to add a liquid interface for config fetching directly to the master, one that can also discern the minion loading the configs based on targets.
But I do agree with @cachedout , this is well written , properly implemented, and very well thought out.

**Review by thatch45 — APPROVED** [review]

File: salt/modules/saltutil.py (hunk 1 above)
Comment (rallytime) [reply salt/modules/saltutil.py]: This won't make it into Oxygen as we've already branched after the feature freeze. So this will need to change to `Fluorine`.

**Comment by mirceaulinic** [comment]
Thank you @thatch45 and @cachedout for the appreciation. As you have approved this PoC, I will go ahead and continue working on this, make it more stable, write tests & documentation and other things needed for this to work fine. I hope to finish this till Fluorine. :-)

@cachedout @rallytime would you prefer me to close this PR, and reopen when ready, or better to just push more commits in this branch and let you know when ready? Thanks!

**Comment by rallytime** [comment]
@mirceaulinic Whatever is easier for you is just fine. :)

**Comment by mirceaulinic** [comment]
Alright, then I will leave this open and let you know when it's ready for re-review. Thanks!

**Comment by isbm** [comment]
I was away on vacations. Reading over this, seems like I am probably going to write something opposite from :heart: and :fireworks: ... Please do not get me wrong: I am only trying to lay down here my concerns. As long as we have answers on that, I am fine.

While storing the configuration inside Mongo is asking for a serious trouble, I would skip this topic, because any module could store anywhere else than just Mongo, that's fine. But what worries me is this:

- Centralised point for the configuration. How that supposed to work offline, once maintenance window cuts the network and scheduler should so something at that time? Seems like this solution allows _either_ way. But once you have hybrid config, I hardly see the point of even having all that. 

- Centralised point for the configuration is, again, _elsewhere_. Why not just taking the whole config from the master instead? Having a plugin system implies "many places, depends on the plugin". But... why? Usually, when we have options, means we don't know exactly what we want. Instead, we could store configs actually on the master in some sort of repo (GitFS, perhaps?) and make it available to the minions, once we want it centralised and always online.

- I am not really sure that regenerating config is that problem. More interesting would be to hot-update config, so the minion does not need to be restarted, only config is re-read. Taking to the account that config is being read up to 200+ times (I once calculated it in various conditions), probably this would be the first priority to fix, before we taking it elsewhere.

- Debugging issues. If you have a minion and you want to snapshot the configuration, so far just copy files and you're done. Now, with this approach, if a module does not allows you too export config, you're in a trouble. And since there is no policy/rule/interface for it, we will have a rock-n-roll pretty soon with this, once people will start developing "simple modules" for this. And again, this adds complexity and debugging struggle, while the value of this is pretty low.

- Ironically, it still requires local configuration. So now you have config here, config there. This adds more complexity instead of simplify the thing. With my last "invention" of SSDP automatic Master recognition, we could instead get rid of config entirely, if you want so. Then Master would just return a Minion config options on its boot-up. And the configs would be on Master, some sort of common-shared between Masters (if a cluster of those). In this case we do not need arbitrary modular stuff that can result storing config somewhere unexpected, but would allow us to get the goal of having things centralised, if one would need it so. I would do something that direction instead.

- The integration issue. It would be not exactly my favourite approach to update some database from other app where Minion's config is stored. Instead, I would like to have API to the Master, and knock there, saying "please set Foo to X on minion/group-of-minions Z". That would be better isolation, I'd say. In this case again, we need no modules as Master is storing the config values (or at least through the Master).

- At SUSE we already using this sort of centralised config store over GitFS in our infra.

As I understand correctly (or not, please then fix me), the whole idea is to centralise configuration so having some sort of groups, where we could just have a config-per-a-group. So all we need to do is to change a config of some specific group and have this pushed everywhere. Like I said, GitFS can do this already. Or even some common network share (not entirely nice, but still the same effect).

The bottom line, files aren't always bad. And grains, job IDs, serialised objects — all that isn't in the database either, but just a files. Why adding such overhead to the configuration then?

I don't know if this makes sense to you completely :wink: but maybe you may find something of this helpful.

**Review by isbm — COMMENTED** [review]
See my big comment...

**Comment by mirceaulinic** [comment]
Hi @isbm: thank you for taking your time to put all these together - very good points!

> Centralised point for the configuration. How that supposed to work offline, once maintenance window cuts the network and scheduler should do something at that time?

That's right, and it would be the equivalent of the external pillars: if the system is offline, no more pillars available. Ideally the external opts should be loaded only on startup (as in opposite to being able to refresh them during runtime - like the pillars), and that links to another point you made.

> Usually, when we have options, means we don't know exactly what we want. Instead, we could store configs actually on the master in some sort of repo (GitFS, perhaps?) and make it available to the minions, once we want it centralised and always online.

Yes, I agree with you. But one of the things I like the most in Salt is that you don't have hard constraints like that, but it's flexible enough to achieve whatever you want. And my idea is in this direction (at least that's what I aim to).

> Taking to the account that config is being read up to 200+ times (I once calculated it in various conditions), probably this would be the first priority to fix, before we taking it elsewhere.

+1 we should tackle this problem indeed. Would you have anywhere more notes to dive into this subject? I also noticed that the config files are read several times, but I wasn't aware about the exact number.

> Debugging issues.

Can't agree more: it can become an excellent gun to shoot yourself in the foot. But again, external pillars have the same downside too.

> Now, with my last "invention" of SSDP automatic Master recognition, Master is able to return any data to the Minion. 

That sounds nice. Is this going to be available in Oxygen?

> I would be happy if some of configs could be inherited, like "common config" that applies to multiple configs and specific config that applies only to the specific group or even machine.

This is a feature I would like achieve, i.e., the possibility to have some options in common (some really need to be in sync), and others that are specific. From my rudimentary example above, I tried to show this:

```/etc/salt/master```:
```yaml
liquid:
  - mongo:
      host: example.com
      port: 49017
      db: common_opts
  - mongo:
      host: example.com
      port: 49017
      db: master_only_opts
```

```/etc/salt/minion```
```yaml
liquid:
  - mongo:
      host: example.com
      port: 49017
      db: common_opts
  - mongo:
      host: example.com
      port: 49017
      db: minion_only_opts
```

Is this what you meant?

> As I understand correctly (or not, please then fix me), the whole idea is to centralise configuration so having some sort of groups, where we could just have a config-per-a-group. So all we need to do is to change a config of some specific group and have this pushed everywhere.

Correct.

> Like I said, GitFS can do this already. 

Can't agree more, but it's only GitFS. And going back to external pillars: while external pillars can fetch data from other sources including GitFS, why couldn't we do the same with the opts.

> The bottom line, files aren't always bad.

Certainly.

> Why adding such overhead to the configuration then?

The way I look at it, it's not an overhead but a flexibility similar to the external pillars, external tops, external caches, etc. It's not to be used by everyone, and nobody would be constrained to use it: but if you want it, it's there. :-)

Disclaimer: I am not 100% convinced this is the best thing we can do about the config opts, but I didn't hear any strong argument against. But I agree there are some other (more important) problems to be solved beforehand.

I am looking forward to debate this further! Cheers & thanks for your thoughts.

**Comment by damon-atkins** [comment]
One options is to keep the minion simple. e.g. 
`saltconf://servername ` .
or
`salt://_minion_/__default__` (large list of options) followed by  `salt://_minion_/servername` (small specific list of options) which updates any settings included in default.  

How the master delivers this could be a flat file system or gitfs or helper etc.
It's a basic change to the minion.

The master can down, or even dns broken (because the dns server is down or resolv.conf stuffed) and hence the master should fullback to a last used IP address if it can not do a name look up. Then you can stil fix resolv.conf

**Comment by isbm** [comment]
@mirceaulinic in my eyes, external **pillars** concept while similar, still cannot be compared here with the **configuration** because of the very paradigm:
- they are _optional_ while config is _mandatory_
- they make more sense to be elsewhere, as they carry sensitive data and so such data might be protected under different required policies
- they usually _do not belong_ to the Salt infrastructure but carry the "executive" information elsewhere.
- they are meant to integrate processes within the infrastructure, while config is meant to define components in the infrastructure 
- they are non-debuggable in the sense that they are not vital for the basic minion functionality
- they are even conceptually differently debuggable: it boils down to the value appears in templates/SLS/results or not, while the configuration directly "hammering" the internal stats of the Minion, and thus no transparency here overcomplicates everything instead
- the beacon configuration in them usually is a trouble to debug
- we do both ext pillars and config via GitFS. While it is sort of "remote", it is still local in temp directory.

Because of these distinctions I tend to disagree that external pillars can be comparable to the configuration, although _technically_ you are right.

Why I think plugin system is not needed, because files are faster and more reliable. Storing them on the Master — yes. Storing the data elsewhere — you sure? I can already give you a list of horror scenarios, where your idea will bring a problems: you store that in Mongo on some node on some AWS. This goes down, you cannot contact the admin of this; or you store that on GitHub and it goes down and you are in a trouble etc. But if you have your physical files — they are fast and reliable, but network gets slow or broken and your 10K machines can have a rock-n-roll.

Summarising, I find your idea a really good start, because it is about **centralising the configuration** which is certainly missing in Salt. But I would be very careful here and so would not overdo this either, just because we can. **Unlike pillars, since config belongs to the Salt infra, I would keep it there and nowhere else.** What we could do is this:

- Configs are read multiple times right now. This needs to be fixed _first_.
- Configs should be manage-able at Master side. APIs probably could only say "set X to key Z on minion Foo" and that's it.
- Hot-config of the Minion (comes after we fix the first problem of re-reading them everywhere)
- SSDP feature could additionally expose them to the Minions, once used. (there is a huge hall of innovations and improvements!) and I would look how to operate configs from the master, keeping them exactly there. But I would not really overdo this, because if there is a way to do it wrong, you will get this most of the time.

Answering your questions:

- yes, SSDP it is already in Oxygen. It is even turned ON by default and will yank your minion for 20 seconds on startup if you configure it wrong. :wink (See: https://github.com/saltstack/salt/issues/45056)
- yes, I tried to keep all opts in the dunder single variable and singleton even. Doesn't completely work: the cloud part was still very messy and was failing many integration tests. The `opts` in Master and Minion classes gets different I abandoned that for now, unfortunately. But I would happily jump in on this again!

**Comment by isbm** [comment]
@mirceaulinic a bit more insight to the config reread last time I did. Basically, there is kinda "get/apply" config in the `config/__init__.py` stuff which is called right everywhere, sometimes in a for cycles (maybe not anymore, but was the case for 2015.8, AFAIK. So what I did is something like (pseudocode):
```bash
function make_config_for_something():
    if __cached_config__ is empty then:
        __cached_config__.update(read_all_possible_configs())
    end
    return __cached_config__.get_content()
```

There are floating several copies in the memory of the `__opts__`. Certainly one that is `self.opts` of Master/Minion/Syndic classes has different `id` than the originally passed parameter to the constructor, which means we at least have two copies of those all the time (it is still the case in the Oxygen right now, just checked). My attempt to fix it the way above wasn't completely successful (although it significantly lowered down startup time and memory usage was also noticeably smaller). Still something wasn't working well for cloud part and tests were failing. But then I didn't dived into that anymore.

**Comment by damon-atkins** [comment]
It's a pity that salt does not start a cache process e.g for config files, grain results, etc , at least this way it a new process would fetch the data from the cache process instead of the file system.

**Comment by isbm** [comment]
@damon-atkins with your PR it could! :wink:

**Comment by rallytime** [comment]
@mirceaulinic We recently merged a PR that is the beginning of an RFC process. The main docs for that are located [here](https://github.com/saltstack/salt/blob/develop/rfcs/README.md). It would be wonderful if you had a chance to review that document and see how this PR/idea would fit in with that. (I realize you have already written some code here, but there is some excellent discussion from you and @isbm.) 

After discussing the RFC process with @saltstack/team-suse, this PR came up so I wanted to mention it here. We're going to start moving in this direction to use this kind of process moving forward and since you've added quite a few excellent features in Salt, you're feedback would be greatly appreciated.

**Comment by isbm** [comment]
Yes. I would definitely put all that in an RFC and let's shape this out together. Because I see a lot of nice things here from @mirceaulinic but I also see a lot of danger and just [YAGNI](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it).

**Comment by mirceaulinic** [comment]
The RFC process is a great idea. I agree @rallytime @isbm I will submit an RFC soon, and we'll discuss this further. Thanks!
